src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/ExportProviderBuilder.cs

```diff
@@ -59,20 +105,52 @@ public static async Task<ExportProvider> CreateExportProviderAsync(
+        if (compositionCacheFile is not null)
+        {
+            if (Path.GetDirectoryName(compositionCacheFile) is string directory)
+            {
+                Directory.CreateDirectory(directory);
+            }
 
-        // Create an export provider, which represents a unique container of values.
-        // You can create as many of these as you want, but typically an app needs just one.
-        var exportProvider = exportProviderFactory.CreateExportProvider();
+            CachedComposition cachedComposition = new();
+            var tempFilePath = Path.Combine(Path.GetTempPath(), Path.GetTempFileName());
+            using (FileStream cacheStream = new(tempFilePath, FileMode.Create, FileAccess.Write, FileShare.None, 4096, useAsync: true))
+            {
+                await cachedComposition.SaveAsync(config, cacheStream);
+            }
 
-        // Immediately set the logger factory, so that way it'll be available for the rest of the composition
-        exportProvider.GetExportedValue<ServerLoggerFactory>().SetFactory(loggerFactory);
+            File.Move(tempFilePath, compositionCacheFile, overwrite: true);
+        }
 
-        // Also add the ExtensionAssemblyManager so it will be available for the rest of the composition.
-        exportProvider.GetExportedValue<ExtensionAssemblyManagerMefProvider>().SetMefExtensionAssemblyManager(extensionManager);
+        // Prepare an ExportProvider factory based on this graph.
+        return config.CreateExportProviderFactory();
+    }
 
-        return exportProvider;
+    private static string GetCompositionCacheFilePath(string cacheDirectory, ImmutableArray<string> assemblyPaths)
+    {
+        // Include the .NET runtime version in the cache path so that running on a newer
+        // runtime causes the cache to be rebuilt.
+        var cacheSubdirectory = $".NET {Environment.Version.Major}";
+        return Path.Combine(cacheDirectory, cacheSubdirectory, $"c#-languageserver.{ComputeAssemblyHash(assemblyPaths)}.mef-composition");
+
+        static string ComputeAssemblyHash(ImmutableArray<string> assemblyPaths)
+        {
+            var assemblies = new StringBuilder();
+            foreach (var assemblyPath in assemblyPaths)
+            {
+                // Include assembly path in the hash so that changes to the set of included
+                // assemblies cause the composition to be rebuilt.
+                assemblies.Append(assemblyPath);
+                // Include the last write time in the hash so that newer assemblies written
+                // to the same location cause the composition to be rebuilt.
+                assemblies.Append(File.GetLastWriteTimeUtc(assemblyPath).ToString("F"));
+            }
+
+            var hash = SHA256.HashData(Encoding.UTF8.GetBytes(assemblies.ToString()));
+            // Convert to filename safe base64 string.
+            return Convert.ToBase64String(hash).Replace('+', '-').Replace('/', '_').TrimEnd('=');
```

Comment:
>Replace [](http://example.com/codeflow?start=48&length=7)

we're pretty confident that these are the only invalid chars? If not, would it be safer to use Path.GetInvalidPathChars and convert those to underscores?

Reply:
According to the base64 spec those are the trouble characters. We could bring in the Base64Url by adding a package reference ([docs](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.text.base64url?view=net-9.0). Once we move to .net 9.0 we can drop the package as it will be built-in.

Reply:
Interesting, I hadn't realized that + and / were the only non-alphanumeric values in base64. Out of curiosity, which OS has + as a problematic file path char?

Reply:
I am not sure. I was trying to mimic [RC 3548](https://www.rfc-editor.org/rfc/rfc3548#page-6) which is why I made those two substitutions.